ICMP Message specification

Specify all ICMP messages in the ICMP packet.
Relevant RFCs:
 - [RFC 792](https://tools.ietf.org/html/rfc792)
 - [RFC 1812](https://tools.ietf.org/html/rfc1812)
 - [RFC 4884](https://tools.ietf.org/html/rfc4884)

I tried to express all ICMP variants in one message but clearly i am doing something wrong. I get this error message but do not really understand it (especially the last part). Could someone have a look at this please? I added a commit with the "new" Spec to branch 174.
```
rflx.model.ModelError: contradicting condition 1 from field "Checksum" to "Identifier" on path [Tag -> Code_Time_Exceeded -> Checksum] in "ICMP.ICMP_Message" (Tag = Time_Exceeded
E                          ∧ Echo_Request = 8
E                          ∧ ((((Tag = Echo_Request
E                           or Tag = Echo_Reply)
E                          or Tag = Timestamp_Msg)
E                         or Tag = Timestamp_Reply)
E                        or Tag = Information_Request)
E                       or Tag = Information_Reply
E                          ∧ Time_Exceeded = 11
E                          ∧ Information_Request = 15
E                          ∧ Echo_Reply = 0
E                          ∧ Information_Reply = 16
E                          ∧ Timestamp_Reply = 14
E                          ∧ Timestamp_Msg = 13)
```

--- a/examples/icmp_socket.py
+++ b/examples/icmp_socket.py
@@ -34,9 +34,9 @@ class ICMPSocket:
             print("ICMP data is equal")
 
     def __create_msg(self) -> MessageValue:
-        icmp = self.package_icmp["Echo_Request_Reply_Message"]
+        icmp = self.package_icmp["Message"]
         icmp.set("Tag", "Echo_Request")
-        icmp.set("Code", 0)
+        icmp.set("Code_Zero", 0)
         icmp.set("Checksum", 12824)
         icmp.set("Identifier", 5)
         icmp.set("Sequence_Number", 1)

--- a/rflx/pyrflx/typevalue.py
+++ b/rflx/pyrflx/typevalue.py
@@ -480,6 +480,8 @@ class MessageValue(TypeValue):
         self._fields[INITIAL.name] = initial
         self._simplified_mapping: Mapping[Name, Expr] = {}
         self._preset_fields(INITIAL.name)
+        self.accessible_fields: List[str]
+        self._update_accessible_fields()
 
     def __copy__(self) -> "MessageValue":
         return MessageValue(self._type, self._refinements)
@@ -515,17 +517,28 @@ class MessageValue(TypeValue):
     def _prev_field(self, fld: str) -> str:
         if fld == INITIAL.name:
             return ""
+        prev: List[str] = []
         for l in self._type.incoming(Field(fld)):
             if self.__simplified(l.condition) == TRUE:
-                return l.source.name
+                prev.append(l.source.name)
+
+        if len(prev) == 1:
+            return prev[0]
+        for field in prev:
+            if field in self.accessible_fields:
+                return field
         return ""
 
     def _get_length_unchecked(self, fld: str) -> Expr:
-        for l in self._type.incoming(Field(fld)):
-            if self.__simplified(l.condition) == TRUE and l.length != UNDEFINED:
-                return self.__simplified(l.length)
-
         typeval = self._fields[fld].typeval
+        if isinstance(typeval, CompositeValue):
+            for l in self._type.incoming(Field(fld)):
+                if (
+                    self.__simplified(l.condition) == TRUE
+                    and l.length != UNDEFINED
+                    and self._fields[l.source.name].set
+                ):
+                    return self.__simplified(l.length)
         if isinstance(typeval, ScalarValue):
             return typeval.size
         return UNDEFINED
@@ -699,6 +712,7 @@ class MessageValue(TypeValue):
             )
 
         self._preset_fields(field_name)
+        self._update_accessible_fields()
 
     def _preset_fields(self, fld: str) -> None:
         nxt = self._next_field(fld)
@@ -762,8 +776,7 @@ class MessageValue(TypeValue):
     def fields(self) -> List[str]:
         return [f.name for f in self._type.fields]
 
-    @property
-    def accessible_fields(self) -> List[str]:
+    def _update_accessible_fields(self) -> None:
         nxt = self._next_field(INITIAL.name)
         fields: List[str] = []
         while nxt and nxt != FINAL.name:
@@ -781,7 +794,7 @@ class MessageValue(TypeValue):
 
             fields.append(nxt)
             nxt = self._next_field(nxt)
-        return fields
+        self.accessible_fields = fields
 
     def _is_valid_opaque_field(self, field: str) -> bool:
         if self._get_length_unchecked(field) == UNDEFINED:

--- a/specs/icmp.rflx
+++ b/specs/icmp.rflx
@@ -41,97 +41,75 @@ package ICMP is
    type Sequence_Number is mod 2**16;
    type Pointer is mod 2**8;
    type Timestamp is mod 2**32;
-   type Gateway_Inet_Addr is mod 2**32;
-   type Uint_32_Unused is range 0 .. 0 with Size => 32;
-   type Uint_24_Unused is range 0 .. 0 with Size => 24;
+   type Gateway_Internet_Address is mod 2**32;
+   type Unused_32 is range 0 .. 0 with Size => 32;
+   type Unused_24 is range 0 .. 0 with Size => 24;
 
-   type Echo_Request_Reply_Message is
+   type Message is
       message
-         Tag : Tag;
-         Code : Code_Zero;
-         Checksum : Checksum;
+         Tag : Tag
+            then Code_Destination_Unreachable
+               if Tag = Destination_Unreachable,
+            then Code_Time_Exceeded
+               if Tag = Time_Exceeded,
+            then Code_Redirect
+               if Tag = Redirect,
+            then Code_Zero
+               if (Tag = Echo_Reply
+                   or Tag = Echo_Request
+                   or Tag = Source_Quench
+                   or Tag = Parameter_Problem
+                   or Tag = Timestamp_Msg
+                   or Tag = Timestamp_Reply
+                   or Tag = Information_Request
+                   or Tag = Information_Reply);
+         Code_Destination_Unreachable : Code_Destination_Unreachable
+            then Checksum;
+         Code_Time_Exceeded : Code_Time_Exceeded
+            then Checksum;
+         Code_Redirect : Code_Redirect
+            then Checksum;
+         Code_Zero : Code_Zero;
+         Checksum : Checksum
+            then Unused_32
+               if (Tag = Destination_Unreachable
+                   or Tag = Source_Quench
+                   or Tag = Time_Exceeded),
+            then Identifier
+               if (Tag = Echo_Request
+                   or Tag = Echo_Reply
+                   or Tag = Timestamp_Msg
+                   or Tag = Timestamp_Reply
+                   or Tag = Information_Request
+                   or Tag = Information_Reply),
+            then Pointer
+               if Tag = Parameter_Problem,
+            then Gateway_Internet_Address
+               if Tag = Redirect;
+         Gateway_Internet_Address : Gateway_Internet_Address
+            then Data
+               with Length => 224;
+         Pointer : Pointer;
+         Unused_24 : Unused_24
+            then Data
+               with Length => 224;
+         Unused_32 : Unused_32
+            then Data
+               with Length => 224;
          Identifier : Identifier;
          Sequence_Number : Sequence_Number
+            then Originate_Timestamp
+               if Tag = Timestamp_Msg or Tag = Timestamp_Reply,
+            then null
+               if Tag = Information_Request or Tag = Information_Reply,
             then Data
-               with Length => Message'Last - Sequence_Number'Last;
-         Data : Opaque;
-      end message;
-
-   type Destination_Unreachable_Message is
-      message
-         Tag : Tag;
-         Code : Code_Destination_Unreachable;
-         Checksum : Checksum;
-         Unused : Uint_32_Unused
-            then Data
-               with Length => 224;
-         Data : Opaque;
-      end message;
-
-   type Time_Exceeded_Message is
-      message
-         Tag : Tag;
-         Code : Code_Time_Exceeded;
-         Checksum : Checksum;
-         Unused : Uint_32_Unused
-            then Data
-               with Length => 224;
-         Data : Opaque;
-      end message;
-
-   type Parameter_Problem_Message is
-      message
-         Tag : Tag;
-         Code : Code_Zero;
-         Checksum : Checksum;
-         Pointer : Pointer;
-         Unused : Uint_24_Unused
-            then Data
-               with Length => 224;
-            Data : Opaque;
-      end message;
-
-   type Source_Quench_Message is
-      message
-         Tag : Tag;
-         Code : Code_Zero;
-         Checksum : Checksum;
-         Unused : Uint_32_Unused
-            then Data
-               with Length => 224;
-         Data : Opaque;
-      end message;
-
-   type Redirect_Message is
-      message
-         Tag : Tag;
-         Code : Code_Redirect;
-         Checksum : Checksum;
-         Gateway_Inet_Addr : Gateway_Inet_Addr
-            then Data
-               with Length => 224;
-         Data : Opaque;
-      end message;
-
-   type Timestamp_Request_Reply_Message is
-      message
-         Tag : Tag;
-         Code : Code_Zero;
-         Checksum : Checksum;
-         Identifier : Identifier;
-         Sequence_Number : Sequence_Number;
+               with Length => Message'Last - Sequence_Number'Last
+               if Tag = Echo_Reply or Tag = Echo_Request;
          Originate_Timestamp : Timestamp;
          Receive_Timestamp : Timestamp;
-         Transmit_Timestamp : Timestamp;
-      end message;
-
-   type Information_Request_Reply_Message is
-      message
-         Tag : Tag;
-         Code : Code_Zero;
-         Checksum : Checksum;
-         Identifier : Identifier;
-         Sequence_Number : Sequence_Number;
+         Transmit_Timestamp : Timestamp
+            then null;
+         Data : Opaque;
       end message;
 
 end ICMP;

--- a/tests/test_pyrflx.py
+++ b/tests/test_pyrflx.py
@@ -138,9 +138,9 @@ def fixture_alert(tls_alert_package: Package) -> MessageValue:
     return tls_alert_package["Alert"]
 
 
-@pytest.fixture(name="echo_request_reply_message")
+@pytest.fixture(name="icmp")
 def fixture_echo_request_reply_message(icmp_package: Package) -> MessageValue:
-    return icmp_package["Echo_Request_Reply_Message"]
+    return icmp_package["Message"]
 
 
 @pytest.fixture(name="message_odd_length")
@@ -583,23 +583,23 @@ def test_tls_length_unchecked(tls_record: MessageValue) -> None:
     assert not isinstance(tls_record._get_length_unchecked("Fragment"), Number)
 
 
-def test_icmp_echo_request(echo_request_reply_message: MessageValue) -> None:
+def test_icmp_echo_request(icmp: MessageValue) -> None:
     test_data = (
         b"\x4a\xfc\x0d\x00\x00\x00\x00\x00\x10\x11\x12\x13\x14\x15\x16\x17"
         b"\x18\x19\x1a\x1b\x1c\x1d\x1e\x1f\x20\x21\x22\x23\x24\x25\x26\x27"
         b"\x28\x29\x2a\x2b\x2c\x2d\x2e\x2f\x30\x31\x32\x33\x34\x35\x36\x37"
     )
 
-    echo_request_reply_message.set("Tag", "Echo_Request")
-    echo_request_reply_message.set("Code", 0)
-    echo_request_reply_message.set("Checksum", 12824)
-    echo_request_reply_message.set("Identifier", 5)
-    echo_request_reply_message.set("Sequence_Number", 1)
-    echo_request_reply_message.set(
+    icmp.set("Tag", "Echo_Request")
+    icmp.set("Code_Zero", 0)
+    icmp.set("Checksum", 12824)
+    icmp.set("Identifier", 5)
+    icmp.set("Sequence_Number", 1)
+    icmp.set(
         "Data", test_data,
     )
-    assert echo_request_reply_message.bytestring == b"\x08\x00\x32\x18\x00\x05\x00\x01" + test_data
-    assert echo_request_reply_message.valid_message
+    assert icmp.bytestring == b"\x08\x00\x32\x18\x00\x05\x00\x01" + test_data
+    assert icmp.valid_message
 
 
 def test_value_mod() -> None:
@@ -753,18 +753,18 @@ def test_check_nodes_opaque(tlv_checksum: MessageValue, frame: MessageValue) -> 
     assert not frame._is_valid_opaque_field("Payload")
 
 
-def test_icmp_parse_binary(echo_request_reply_message: MessageValue) -> None:
+def test_icmp_parse_binary(icmp: MessageValue) -> None:
     test_bytes = (
         b"\x08\x00\xe1\x1e\x00\x11\x00\x01\x4a\xfc\x0d\x00\x00\x00\x00\x00"
         b"\x10\x11\x12\x13\x14\x15\x16\x17\x18\x19\x1a\x1b\x1c\x1d\x1e\x1f"
         b"\x20\x21\x22\x23\x24\x25\x26\x27\x28\x29\x2a\x2b\x2c\x2d\x2e\x2f"
         b"\x30\x31\x32\x33\x34\x35\x36\x37"
     )
-    echo_request_reply_message.parse(test_bytes)
-    assert echo_request_reply_message.valid_message
-    assert echo_request_reply_message.bytestring == test_bytes
-    assert echo_request_reply_message.accepted_type == bytes
-    assert echo_request_reply_message.size == Number(448)
+    icmp.parse(test_bytes)
+    assert icmp.valid_message
+    assert icmp.bytestring == test_bytes
+    assert icmp.accepted_type == bytes
+    assert icmp.size == Number(448)
 
 
 def test_ethernet_parse_binary(frame: MessageValue) -> None:
@@ -898,8 +898,8 @@ def test_array_assign_incorrect_values(
     tlv.set("Length", 4)
     tlv.set("Value", b"\x00\x00\x00\x00")
 
-    frame.set("Source", 0)
     frame.set("Destination", 0)
+    frame.set("Source", 0)
     frame.set("Type_Length_TPID", 47)
     frame.set("Type_Length", 1537)
     frame.set("Payload", bytes(46))

